File: bowling.go
package bowling

// NewGame is the Game constructor.
func NewGame() Game {
	return Game{}
}

// Game is the main type for the bowling game.
type Game struct {
	currentRoll int
	rolls       [21]int
}

// Roll takes the number of pins knocked over in a turn.
func (g *Game) Roll(p int) {
	g.rolls[g.currentRoll] = p
	g.currentRoll++
}

// Score calculates the score from the current state.
func (g *Game) Score() int {
	var (
		score, frameIdx int
	)
	for frame := 0; frame < 10; frame++ {
		if g.isStrike(frameIdx) {
			score += 10 + g.strikeBonus(frameIdx)
			frameIdx++
			continue
		}
		if g.isSpare(frameIdx) {
			score += 10 + g.rolls[frameIdx+2]
			frameIdx += 2
			continue
		}
		score += g.sumOfBallsInFrame(frameIdx)
		frameIdx += 2
	}
	return score
}

func (g *Game) strikeBonus(i int) int {
	return g.rolls[i+1] + g.rolls[i+2]
}

func (g *Game) sumOfBallsInFrame(i int) int {
	return g.rolls[i] + g.rolls[i+1]
}

func (g *Game) isStrike(i int) bool {
	return g.rolls[i] == 10
}

func (g *Game) isSpare(i int) bool {
	return g.rolls[i]+g.rolls[i+1] == 10
}

File: ruby/lib/bowling.rb
require "bowling/version"

module Bowling
  class Game
    def initialize()
      @current_roll = 0
      @rolls = Array.new(21, 0)
    end

    def roll(pins)
      @rolls[@current_roll] = pins
      @current_roll += 1
    end

    def score
      score, frame, frame_idx = 0, 0, 0
      while frame < 10
        if is_strike frame_idx
          score += 10 + strike_bonus(frame_idx)
          frame_idx += 1
        elsif is_spare(frame_idx)
          score += 10 + spare_bonus(frame_idx)
          frame_idx += 2
        else
          score += sum_of_balls_in_frame(frame_idx)
          frame_idx += 2
        end

        frame += 1
      end
      score
    end

    def is_spare(idx)
      @rolls[idx] + @rolls[idx + 1] == 10
    end

    def is_strike(idx)
      @rolls[idx] == 10
    end

    def spare_bonus(idx)
      @rolls[idx + 2]
    end

    def strike_bonus(idx)
      @rolls[idx + 1] + @rolls[idx + 2]
    end

    def sum_of_balls_in_frame(idx)
      @rolls[idx] + @rolls[idx + 1]
    end
  end
end
